server/src/main/scala/org/http4s/server/armeria/ArmeriaServerBuilder.scala

```diff
@@ -0,0 +1,334 @@
+    armeriaServerBuilder.tls(key, keyCertChain: _*)
+    this
+  }
+
+  /** Configures SSL or TLS of this [[BackendServer]] with the specified [[KeyManagerFactory]].
+    *
+    * @see [[withTlsCustomizer(scala.Function1)]]
+    */
+  def withTls(keyManagerFactory: KeyManagerFactory): Self = {
+    armeriaServerBuilder.tls(keyManagerFactory)
+    this
+  }
+
+  /** Configures SSL or TLS of the [[BackendServer]] with an auto-generated self-signed certificate.
+    * '''Note:''' You should never use this in production but only for a testing purpose.
+    *
+    * @see [[withTlsCustomizer(scala.Function1)]]
+    */
+  def withTlsSelfSigned: Self = {
+    armeriaServerBuilder.tlsSelfSigned
+    this
+  }
+
+  /** Adds the specified `tlsCustomizer` which can arbitrarily configure the [[SslContextBuilder]] that will be
+    * applied to the SSL session.
+    */
+  def withTlsCustomizer(tlsCustomizer: SslContextBuilder => Unit): Self = {
+    armeriaServerBuilder.tlsCustomizer(ctxBuilder => tlsCustomizer(ctxBuilder))
+    this
+  }
+
+  /** Sets the amount of time to wait after calling [[BackendServer#stop()]] for
+    * requests to go away before actually shutting down.
+    *
+    * @param quietPeriod the number of milliseconds to wait for active
+    *                    requests to go end before shutting down. [[Duration.Zero]] means
+    *                    the server will stop right away without waiting.
+    * @param timeout     the amount of time to wait before shutting down the server regardless of active
+    *                    requests.
+    *                    This should be set to a time greater than `quietPeriod` to ensure the server
+    *                    shuts down even if there is a stuck request.
+    */
+  def withGracefulShutdownTimeout(quietPeriod: Duration, timeout: Duration): Self = {
+    armeriaServerBuilder.gracefulShutdownTimeoutMillis(quietPeriod.toMillis, timeout.toMillis)
+    this
+  }
+
+  /** Sets the [[MeterRegistry]] that collects various stats. */
+  def withMeterRegistry(meterRegistry: MeterRegistry): Self = {
+    armeriaServerBuilder.meterRegistry(meterRegistry)
+    this
+  }
+
+  private def shutdown(armeriaServer: BackendServer): F[Unit] =
+    F.async[Unit] { cb =>
+      val _ = armeriaServer
+        .stop()
+        .whenComplete { (_, cause) =>
+          if (cause == null)
+            cb(Right(()))
+          else
+            cb(Left(cause))
+        }
+    }
+
+  override def withBanner(banner: immutable.Seq[String]): Self = copy(banner = banner)
+
+  private def copy(
+      armeriaServerBuilder: ArmeriaBuilder = armeriaServerBuilder,
+      socketAddress: InetSocketAddress = socketAddress,
+      serviceErrorHandler: ServiceErrorHandler[F] = serviceErrorHandler,
+      banner: Seq[String] = banner
+  ): Self =
+    new ArmeriaServerBuilder(armeriaServerBuilder, socketAddress, serviceErrorHandler, banner)
+}
+
+trait ArmeriaServer[F[_]] extends Server[F] {
```

Comment:
I think this could be private[armeria] and give you more bincompat wiggle room.

Reply:
I intentionally made this public to give a chance to access the underlying server. 😀

Reply:
Oops, that's a nuance I didn't catch.  I would probably do that with composition instead of inheritance.  Or maybe we should look at making `Server` a typeclass in http4s-server.  But this is justifiable, too.

Reply:
I added TODO for this. Let me think about this in the following PR.